_plugins/link-checker.rb

```diff
@@ -1,219 +1,376 @@
-        puts "LinkChecker: [Notice] The build will fail if any dead links are found"
-      else
-        puts "LinkChecker: [Notice] The build will fail if a dead internal link is found"
+  def self.verify(_site)
+    return unless @check_links
+
+    @base_url_matcher = %r{^#{@site.config["url"]}#{@site.baseurl}(/.*)$}.freeze
+    retry_hosts = {}
+
+    # Run atleast once
+    loop do
+      urls = @urls
+
+      # If its a retry
+      unless retry_hosts.empty?
+        # Get min sleep time in dict of timeouts and sleep
+        host_name, min_timeout_obj = @retry_timeouts_dict.min_by { |_k, v| v[:retry_timestamp] }
+        now = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+        sleep_time = min_timeout_obj[:retry_timestamp] - now + @@retry_buffer
+
+        if sleep_time > 0
+          Jekyll.logger.info "LinkChecker: [Info] Going to sleep for #{sleep_time}".cyan
+          sleep(sleep_time)
+        end
+
+        # Get URLS to retry and clear from retry hash's
+        Jekyll.logger.info "LinkChecker: [Info] Retrying links for host #{host_name}".cyan
+        urls = retry_hosts[host_name].clone
+        @retry_timeouts_dict.delete(host_name)
+        retry_hosts.delete(host_name)
       end
+
+      # checl each url
+      # - valid URL: should not be failures but or in retry_hosts hash
+      # - invalid URL: should be failures but not in retry_hosts hash
+      # - retry URL: should not be failures, only in retry_hosts hash
+      urls.each do |url, pages|
+        Jekyll.logger.info "LinkChecker: [Info] Checking #{url}".cyan
+
+        valid_or_retry, metadata = check(url)
+        @failures << "Verify:: #{url}, linked to in ./#{pages.to_a.join(', ./')}" unless valid_or_retry
+
+        next unless @retry_external_links and metadata&.key?(:retry_host_name)
+
+        retry_host_name = metadata[:retry_host_name]
+        retry_hosts[retry_host_name] = [] unless retry_hosts.key?(retry_host_name)
+        retry_hosts[retry_host_name] << url
+      end
+
+      @retry_iteration += 1
+      break if !@retry_external_links or (@retry_iteration >= @@max_retry_iterations) or retry_hosts.empty?
     end
 
-    @urls.each do |url, pages|
-      @failures << "#{url}, linked to in ./#{pages.to_a.join(", ./")}" unless self.check(url)
+    unless @failures.empty?
+      msg = "Found #{@failures.size} dead link#{@failures.size > 1 ? 's' : ''}:\n#{@failures.join("\n")}"
     end
-    
-    msg = "Found #{@failures.size} dead link#{@failures.size > 1 ? 's' : ''}:\n#{@failures.join("\n")}" unless @failures.empty?
 
-    if @should_build_fatally
-      raise msg
+    unless retry_hosts.empty?
+      retry_msg = retry_hosts.map do |host, urls|
+        "Host:#{host}\n#{urls.map { |url| "- #{url}" }.join("\n")}\n"
+      end.join("\n")
+      msg = "Links we could not retry: \n#{retry_msg} \n#{msg}"
+    end
+
+    if !@failures.empty?
+      raise msg if @should_build_fatally
+
+      Jekyll.logger.warn "\nLinkChecker: [Warning] #{msg}\n"
```

Comment:
Can we print this in red to attract attention?
```suggestion
      Jekyll.logger.warn "\nLinkChecker: [Warning] #{msg}\n".red
```